File: app/src/main/java/com/appdate/agendamento/adapter/AgendamentoAdapter.java
package com.appdate.agendamento.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.appdate.agendamento.R;
import com.appdate.agendamento.model.Agendamento;
import com.appdate.agendamento.util.DateUtils;
import com.google.firebase.firestore.DocumentChange;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;

import java.util.Calendar;

public class AgendamentoAdapter extends FirestoreAdapter<AgendamentoAdapter.AgendamentoViewHolder> {


    public interface OnAgendamentoSelectedListener {
        void onAgendamentoSelected(DocumentSnapshot restaurant);
    }

    private OnAgendamentoSelectedListener mListener;

    public AgendamentoAdapter(Query query, OnAgendamentoSelectedListener listener) {
        super(query);
        this.mListener = listener;

    }

    @Override
    public int getItemCount() {
        return super.getItemCount();
    }

    @NonNull
    @Override
    public AgendamentoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View v = inflater.inflate(R.layout.card_agendamento, parent, false);

        return new AgendamentoViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull AgendamentoViewHolder holder, int position) {
       holder.setIsRecyclable(false);

       holder.bind(getSnapshot(position), mListener);

    }

       public static class AgendamentoViewHolder extends RecyclerView.ViewHolder {
        TextView txtNome;
        TextView txtProcedimento;
        TextView txtValor;
        TextView txtData;
        TextView txtHora;
        TextView txtDiaDaSemana;

        public AgendamentoViewHolder(View v) {
            super(v);
            txtDiaDaSemana = itemView.findViewById(R.id.txtDiaDaSemana);
            txtNome = itemView.findViewById(R.id.txtNome);
            txtProcedimento = itemView.findViewById(R.id.txtProcedimento);
            txtValor = itemView.findViewById(R.id.txtValor);
            txtData = itemView.findViewById(R.id.txtData);
            txtHora = itemView.findViewById(R.id.txtHora);
        }

        private void bind(final DocumentSnapshot snapshot,
                          final OnAgendamentoSelectedListener listener){
            Agendamento agendamento = snapshot.toObject(Agendamento.class);

            txtNome.setText(agendamento.getNomeCliente());
            txtProcedimento.setText(agendamento.getProcedimento());
            txtValor.setText(itemView.getContext().getString(R.string.adp_valor, agendamento.getValor()));

            String diaDaSemana = DateUtils.diaDaSemana(itemView.getContext(), agendamento.getDataEHora());
            String data = DateUtils.dateMilisToString(agendamento.getDataEHora());
            String dataAtual = DateUtils.dateMilisToString(Calendar.getInstance().getTimeInMillis());
                if(data.trim().equals(dataAtual.trim())){
                    diaDaSemana = "Hoje";
                    txtDiaDaSemana.setAllCaps(true);
                    txtDiaDaSemana.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.colorAccent));

                }
            Log.d("TAG", "");

            txtDiaDaSemana.setText(diaDaSemana);
            txtData.setText(data);
            txtHora.setText(DateUtils.timeMilisToString(agendamento.getDataEHora()));

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null){
                        listener.onAgendamentoSelected(snapshot);
                    }
                }
            });

            }
    }


}

File: app/src/main/java/com/appdate/agendamento/adapter/AgendamentoAdapter.kt
package com.appdate.agendamento.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.appdate.agendamento.R
import com.appdate.agendamento.model.Agendamento
import com.appdate.agendamento.util.DateUtils.dataAtualInMilis
import com.appdate.agendamento.util.DateUtils.dataAtualToString
import com.appdate.agendamento.util.DateUtils.dateMilisToString
import com.appdate.agendamento.util.DateUtils.diaDaSemana
import com.appdate.agendamento.util.DateUtils.stringToDateMilis
import com.appdate.agendamento.util.DateUtils.timeMilisToString
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query

open class AgendamentoAdapter(mQuery: Query?, private var mListener:OnAgendamentoSelectedListener) :
        FirestoreAdapter<RecyclerView.ViewHolder>(mQuery) {

    interface OnAgendamentoSelectedListener {
        fun onAgendamentoSelected(snapshot: DocumentSnapshot?)
    }



    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v: View
        return when (viewType) {
            ListItem.TYPE_GENERAL -> {
                v = inflater.inflate(R.layout.card_agendamento, parent, false)
                AgendamentoViewHolder(v)
            }
            ListItem.TYPE_DATE -> {
                v = inflater.inflate(R.layout.separador_agendamento, parent, false)
                DataViewHolder(v)
            }
            else -> null!!
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        holder.setIsRecyclable(false)
        if (holder is AgendamentoViewHolder) {
            val generalItem = getItem(position) as GeneralItem
            val snapshot:DocumentSnapshot? = generalItem.documentSnapshot
            holder.bind(snapshot, mListener)
        } else if (holder is DataViewHolder) {
            val dateItem = getItem(position) as DateItem
            holder.bind(dateItem)
        }
    }

    class DataViewHolder internal constructor(v: View?) : RecyclerView.ViewHolder(v!!) {
        var txtData: TextView
        var txtDiaDaSemana: TextView
        fun bind(dateItem: DateItem) {
            txtData.text = dateItem.date
            val dataDoSnapshotInMilis = stringToDateMilis(dateItem.date, null)
            var strDiaDaSemana = diaDaSemana(itemView.context, dataDoSnapshotInMilis)
            val dataAtual = dataAtualToString
            val dataDoSnapshot = dateMilisToString(dataDoSnapshotInMilis)
            val diaMilis = 86400000
            val dataDiaSeguinte = dateMilisToString(dataAtualInMilis + diaMilis)
            if (dateItem.date?.trim { it <= ' ' } == dataAtual.trim { it <= ' ' }) {
                strDiaDaSemana = "Hoje"
                txtDiaDaSemana.setTextColor(ContextCompat.getColor(itemView.context, R.color.colorAccent))
                txtData.visibility = View.INVISIBLE
            } else if (dataDoSnapshot.trim { it <= ' ' } == dataDiaSeguinte.trim { it <= ' ' }) {
                strDiaDaSemana = "Amanhã"
                txtDiaDaSemana.setTextColor(ContextCompat.getColor(itemView.context, R.color.corSeparador))
                txtData.visibility = View.INVISIBLE
            }
            txtDiaDaSemana.text = strDiaDaSemana
        }

        init {
            txtData = itemView.findViewById(R.id.txtData)
            txtDiaDaSemana = itemView.findViewById(R.id.txtDiaDaSemana)
        }
    }

    class AgendamentoViewHolder(v: View?) : RecyclerView.ViewHolder(v!!) {
        var txtNome: TextView
        var txtProcedimento: TextView
        var txtValor: TextView
        var txtHora: TextView
        fun bind(snapshot: DocumentSnapshot?,
                 listener: OnAgendamentoSelectedListener?) {
            val agendamento = snapshot?.toObject(Agendamento::class.java)
            txtNome.text = agendamento?.nomeCliente
            txtProcedimento.text = agendamento?.procedimento
            txtValor.text = itemView.context.getString(R.string.adp_valor, agendamento?.valor)
            txtHora.text = timeMilisToString(agendamento?.dataEHora)
            itemView.setOnClickListener { listener?.onAgendamentoSelected(snapshot) }
        }

        init {
            txtNome = itemView.findViewById(R.id.txtNome)
            txtProcedimento = itemView.findViewById(R.id.txtProcedimento)
            txtValor = itemView.findViewById(R.id.txtValor)
            txtHora = itemView.findViewById(R.id.txtHora)
        }
    }

}